qiskit/transpiler/passes/basis/translate_parameterized.py

```diff
@@ -0,0 +1,192 @@
+            supported_gates = target.operation_names
+        elif supported_gates is None:
+            raise ValueError("One of ``supported_gates`` or ``target`` must be specified.")
+
+        self._supported_gates = supported_gates
+        self._target = target
+        self._translator = BasisTranslator(equivalence_library, supported_gates, target=target)
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the transpiler pass.
+
+        Args:
+            dag: The DAG circuit in which the parameterized gates should be unrolled.
+
+        Returns:
+            A DAG where the parameterized gates have been unrolled.
+
+        Raises:
+            QiskitError: If the circuit cannot be unrolled.
+        """
+        qubits_to_indices = {qubit: i for i, qubit in enumerate(dag.qubits)}
+        return self._run(dag, qubits_to_indices)
+
+    def _run(self, dag: DAGCircuit, qubits_to_indices: dict):
+        """Run the decomposition, keeping a global map of qubit to index mappings.
+
+        This is required to map the DAG qubits to indices for the target decomposition.
+        """
+
+        for node in dag.op_nodes():
+            # check whether it is parameterized and we need to decompose it
+            if _is_parameterized(node.op) and not _is_supported(
+                node, self._supported_gates, self._target, qubits_to_indices
+            ):
+                definition = node.op.definition
+
+                if definition is not None:
+                    # recurse to unroll further parameterized blocks
+                    sub_indices = [qubits_to_indices[qubit] for qubit in node.qargs]
+                    sub_dag = circuit_to_dag(definition)
+                    sub_map = dict(zip(sub_dag.qubits, sub_indices))
+                    unrolled = self._run(circuit_to_dag(definition), sub_map)
+                else:
+                    # if we hit a base case, try to translate to the specified basis
+                    try:
+                        unrolled = self._translator.run(_instruction_to_dag(node.op))
+                    except Exception as exc:
+                        raise QiskitError("Failed to translate final block.") from exc
+
+                # replace with unrolled (or translated) dag
+                dag.substitute_node_with_dag(node, unrolled)
+
+        return dag
+
+
+def _is_parameterized(op: Instruction) -> bool:
+    return any(
+        isinstance(param, ParameterExpression) and len(param.parameters) > 0 for param in op.params
+    )
+
+
+def _is_supported(
+    node: DAGOpNode, supported_gates: list[str], target: Target | None, qubits_to_indices: dict
+) -> bool:
+    """Check whether the node is supported.
+
+    If the target is provided, check using the target, otherwise the supported_gates are used.
+    """
+    if target is not None:
+        qubit_indices = [qubits_to_indices[qubit] for qubit in node.qargs]
+        return target.instruction_supported(node.op.name, qubit_indices)
```

Comment:
So I may have given bad advice on this, thinking more critically about this this could be potentially incorrect in the case you're running this pass before a layout is applied. If no layout is applied this pass will treat the qubit position in a virtual circuit as the physical qubit on the target is running on (i.e. q0 in the circuit is q0 on the device) and incorrectly return here. If we can assume the circuit has a layout applied whenever this is run then this is fine (basis translator makes this assumption). But if we can't make this assumption I think dropping the `qubit_indices` here is ok since we're just checking if the target supports the operation broadly and not specifically on any qubit.

In practice all current backends I'm aware of have globally supported gates so this doesn't matter too much, but it's not guaranteed that all gates are available on all qubits in a target.